### DISCUSSION TITLE
Graph Attribute Filter Functions

### DISCUSSION THREAD

**Discussion by gmelasecca** [networkx/networkx#6133] [Q&A]
I am new to NetworkX so please do not scold me for not knowing this. I am working on building a service map for internal use cases where I want to be able to see if nodes have physical connections (phase 1). What I am looking to do or learn about within networkX, is there any functions that allow for attribute filtering?

for instance, if I have node link data such as the following. I want to be able to filter for only nodes or edges where attributes equal a specific condition such as "datacenter == location_2". I know that I can write some python code to parse this data and or the response that NetworkX provides depending on the algorithm or function used but again I am wondering if NetworkX has this kind of feature built in so I dont have to write any wrapper code. 

another quick example: if I were to call all_shortest_paths function and the response is a HUGE list of connections, if I only wanted to see all_shortest_paths for datacenter == location_2. basically offloading the compute to networkX in memory rather then getting the response, then parsing the response, then filtering the response, then creating a new response object. hopefully this helps clarify my ask.

thank you.

    "node_link_data": {
      "directed": "true",
      "graph": {},
      "links": [
        {
          "source": "test_server_1",
          "target": "test_server_2"
        }
      ],
      "multigraph": "false",
      "nodes": [
        {
          "datacenter": "location_2",
          "hello": "world",
          "id": "test_server_3",
          "operational_status": "1",
          "platform": "platform_1",
          "status": "offline"
        },
        {
          "datacenter": "location_2",
          "hello": "world",
          "id": "test_server_2",
          "operational_status": "1",
          "platform": "platform_2",
          "status": "offline"
        },
        {
          "datacenter": "location_1",
          "hello": "world",
          "id": "test_server_1",
          "operational_status": "1",
          "platform": "platform_1",
          "status": "live"
        }
      ]
    }

**Comment by rossbar** [comment]
> is there any functions that allow for attribute filtering?

Yes, filtering on graph, node, and edge attributes is a very common operation in NetworkX! If you're just getting started, you may find the NetworkX tutorial useful, particularly the [section on attributes](https://networkx.org/documentation/latest/tutorial.html#adding-attributes-to-graphs-nodes-and-edges).

To expand on that a bit - the networkx graph objects are extremely flexible, so there are many ways you can add + filter on attributes. For example, if you wanted all the edges in a network that represent connections between *live* servers:

```python
>>> G = nx.DiGraph()
>>> G.add_nodes_from(["server1", "server2", "server3"], status="live")
>>> G.add_edges_from([("server1", "server2"), ("server2", "server3")])

# Extract all connections between live servers from the example network
# Since all servers are live, this will return both edges in the network
>>> [
...     (u, v)
...     for u, v in G.edges
...     if G.nodes[u]["status"] == "live" and G.nodes[v]["status"] == "live"
... ]
>>> [('server1', 'server2'), ('server2', 'server3')]

# Now what happens when server 3 goes down...
>>> G.nodes["server3"]["status"] = "down"
>>> [
...     (u, v)
...     for u, v in G.edges
...     if G.nodes[u]["status"] == "live" and G.nodes[v]["status"] == "live"
... ]
>>> [('server1', 'server2')]
```

There are many other things one could do. For example, the `nx.filters` subpackage contains functionality that will return view/filter objects on the input graph - this may be useful if you have very large graphs and want to save memory.

**Reply by gmelasecca** [reply]
Thank you for the quick and detailed response rossbar. Let me look at the tutorial and mentioned nx.filters above. do you have a link to the nx.filters subpackage? what I found was kind of bare. see here: [networkx_reference.pdf](https://networkx.org/documentation/stable/_downloads/networkx_reference.pdf)

ill give you some context to maybe help elaborate on the initial issue / problem which I am trying to solve. I dont think we have very large graphs, roughly 500k nodes and 1 million edges leys say. Currently we are creating multiple DiGraphs and then combining them into a single graph, lets call it "graph-all". the reason for doing this now is data quality and consistency between multiple sources of data. lets say that source 1 has field values that are datacenter == "12345" and source 2 has field values that are dc_name == "12345". so I am trying to create a filtering logic in my code that is able to accept params such as a dictionary/json object where they can tell me which attributes they want to SEE or NOT see in the response graph object. typically I an responding with a node_link_data object but I am all open to suggestions. 

example params a user or service can pass to my api that will filter the node attributes where the nodes **MUST** contain attribute _"id" == "server_1"_ and attribute "_datacenter_" must exist as an attribute regardless of the value. hopefully this makes more sense what I am looking for. 

while typing this I did find this article as well which I assume is the nx.filter function you mentioned above: [networkx.classes.graphviews.subgraph_view.html](https://networkx.org/documentation/stable/reference/classes/generated/networkx.classes.graphviews.subgraph_view.html )

```
      "must": [
        {
          "match": {
            "id": "server_1.*"
          }
        },
        {
          "exists": {
            "attribute_name": "datacenter"
          }
        }
      ]
```
